core/src/main/java/io/stargate/core/activator/BaseActivator.java

```diff
@@ -0,0 +1,169 @@
+import org.osgi.framework.ServiceEvent;
+import org.osgi.framework.ServiceListener;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class BaseActivator implements BundleActivator, ServiceListener {
+  private static final Logger logger = LoggerFactory.getLogger(BaseActivator.class);
+
+  private final String activatorName;
+
+  private final List<Class<?>> dependentServices;
+
+  private BundleContext context;
+
+  private LinkedHashMap<Class<?>, Object> registeredServices;
+
+  private Class<?> targetServiceClass;
+
+  public boolean started;
+
+  /**
+   * @param activatorName - The name used when logging the progress of registration.
+   * @param dependentServices - List of dependent services that needs to be retrieved using service
+   *     reference.
+   * @param targetServiceClass - This class will be used when registering the service.
+   */
+  public BaseActivator(
+      String activatorName, List<Class<?>> dependentServices, Class<?> targetServiceClass) {
+    this.activatorName = activatorName;
+    this.dependentServices = dependentServices;
+    this.registeredServices = createMapWithNullValues(dependentServices);
+    this.targetServiceClass = targetServiceClass;
+  }
+
+  private LinkedHashMap<Class<?>, Object> createMapWithNullValues(
+      List<Class<?>> dependentServices) {
+    LinkedHashMap<Class<?>, Object> map = new LinkedHashMap<>(dependentServices.size());
+    for (Class<?> dependentService : dependentServices) {
+      map.put(dependentService, null);
+    }
+    return map;
+  }
+
+  /**
+   * When the OSGi invokes the start() method, it will try to get all dependentServices passed to
+   * this class's constructor. It will use the {@link BundleContext#getServiceReference(String)}. If
+   * any of those services are not present, it will register the listeners for them using {@link
+   * BundleContext#addServiceListener(ServiceListener, String)} and not start the target service
+   * (see targetServiceClass). It will wait for a notification denoting that service was registered
+   * using {@link this#serviceChanged(ServiceEvent)}. If all services are present, it will call the
+   * user-provided {@link this#createService(List)} and register it in the OSGi using {@link
+   * BundleContext#registerService(Class, Object, java.util.Dictionary)}.
+   */
+  @Override
+  public synchronized void start(BundleContext context) throws InvalidSyntaxException {
+    logger.info("Starting {} ...", activatorName);
+    this.context = context;
+
+    for (Class<?> dependentService : dependentServices) {
+      ServiceReference<?> serviceReference =
+          context.getServiceReference(dependentService.getName());
+      if (serviceReference == null) {
+        logger.debug(
+            "{} service is null, registering a listener to get notification when it will be ready.",
+            dependentService.getName());
```

Comment:
I guess this may be subject to a (low probability) race since we do not know when exactly the dependent service may be registered.

What do you think about using `org.osgi.util.tracker.ServiceTracker` instead?

Reply:
Very good idea, done.